internal/resolver/dns/dns_resolver_test.go

```diff
@@ -736,12 +749,122 @@ func testDNSResolver(t *testing.T) {
 	}
 }
 
+// DNS Resolver immediately starts polling on an error from grpc. This should continue until the ClientConn doesn't
+// send back an error from updating the DNS Resolver's state.
+func TestDNSResolverExponentialBackoff(t *testing.T) {
+	defer leakcheck.Check(t)
+	nt := newTimer
+	defer func(func(d time.Duration) *time.Timer) {
+		newTimer = nt
+	}(nt)
+	timerChan := make(chan *time.Timer, 1)
+	newTimer = func(d time.Duration) *time.Timer {
+		// Will never fire on it's own, allows this test to call timer immediately.
+		t := time.NewTimer(time.Hour)
+		timerChan <- t
+		return t
+	}
+	tests := []struct {
+		target   string
+		addrWant []resolver.Address
+		scWant   string
+	}{
+		{
+			"foo.bar.com",
+			[]resolver.Address{{Addr: "1.2.3.4" + colonDefaultPort}, {Addr: "5.6.7.8" + colonDefaultPort}},
+			generateSC("foo.bar.com"),
+		},
+		{
+			"foo.bar.com:1234",
+			[]resolver.Address{{Addr: "1.2.3.4:1234"}, {Addr: "5.6.7.8:1234"}},
+			generateSC("foo.bar.com"),
+		},
+		{
+			"srv.ipv4.single.fake",
+			[]resolver.Address{{Addr: "2.4.6.8" + colonDefaultPort}},
+			generateSC("srv.ipv4.single.fake"),
+		},
+	}
+	for _, a := range tests {
+		b := NewBuilder()
+		cc := &testClientConn{target: a.target}
+		// Cause ClientConn to return an error.
+		cc.updateStateErr = balancer.ErrBadResolverState
+		r, err := b.Build(resolver.Target{Endpoint: a.target}, cc, resolver.BuildOptions{})
+		timer := <-timerChan
+		timer.Reset(0)
+		if err != nil {
+			t.Fatalf("%v\n", err)
+		}
+		var state resolver.State
+		var cnt int
+		for i := 0; i < 2000; i++ {
+			state, cnt = cc.getState()
+			if cnt > 0 {
+				break
+			}
+			time.Sleep(time.Millisecond)
+		}
+		if cnt == 0 {
+			t.Fatalf("UpdateState not called after 2s; aborting")
+		}
+		if !reflect.DeepEqual(a.addrWant, state.Addresses) {
+			t.Errorf("Resolved addresses of target: %q = %+v, want %+v", a.target, state.Addresses, a.addrWant)
+		}
+		sc := scFromState(state)
+		if a.scWant != sc {
+			t.Errorf("Resolved service config of target: %q = %+v, want %+v", a.target, sc, a.scWant)
+		}
+		// Cause timer to go off 10 times, and see if it calls updateState() correctly.
+		for i := 0; i < 10; i++ {
+			timer := <-timerChan
+			timer.Reset(0)
+			// Propagate.
+			time.Sleep(time.Millisecond)
+		}
+		cc.m1.Lock()
+		if cc.updateStateCalls != 12 {
+			cc.m1.Unlock()
+			t.Fatalf("Exponential backoff is not working as expected - should be updating state at least 10 total times instead of %d", cc.updateStateCalls)
```

Comment:
This should say something like `...should update state 12 times; got %v"` (not "at least 10 total times").

Also: with the lock being taken then given up, is there a chance the printed value is different from the tested value?

Two options:

```go
cc.m1.Lock()
if got := cc.updateStateCalls; got != 12 {
  cc.m1.Unlock() // but see below
  t.Fatalf(..., got)
}
```

or, noting that unlocking before a `Fatal` that kills the `cc.m1` entirely is unnecessary:

```
cc.m1.Lock()
if cc.updateStateCalls != 12 {
  // no Unlock
  t.Fatalf(..., cc.updateStateCalls)
}
```

Or, option 3: do both.

Reply:
Got rid of the unlock() call entirely. I forgot about the read after the unlock. Since the fatal kills mutex anyway, no point of unlocking and persisting got variable. Thanks.

Reply:
Changed the wording of the error from an inequality to an exact number.

Reply:
Btw, just a quick heads up. I had a timer receive earlier in the test that was there due to behavior of making the first Iteration of the watcher loop be determined by the timer. I deleted it, so this became 11 calls expected.

Reply:
Btw, I had a receive from timer channel in test before (which I had due to watcher initially looking up from the first iteration of poll timer). I deleted that, and thus this is actually 11. Just wanted to give you a heads up on next change.